cmd/packaging/windows-msi.go

```diff
@@ -24,26 +31,93 @@ var WindowsMsiTask = &packagingTask{
+		}
+		pngImage, err := png.Decode(f)
+		if err != nil {
+			return "", err
+		}
+		// We can't defer it, because windows reports that the file is used by another program
+		err = f.Close()
+		if err != nil {
+			return "", err
+		}
+		icoImage, err := os.Create(filepath.Join(tmpPath, "build", "assets", "icon.ico"))
+		if err != nil {
+			return "", err
+		}
+		err = ico.Encode(icoImage, pngImage)
 		if err != nil {
 			return "", err
 		}
-		cmdWixl := exec.Command("wixl", "-v", fmt.Sprintf("%s.wxs", packageName), "-o", outputFileName)
-		cmdWixl.Dir = tmpPath
-		cmdWixl.Stdout = os.Stdout
-		cmdWixl.Stderr = os.Stderr
-		err = cmdWixl.Run()
+		// We can't defer it, because windows reports that the file is used by another program
+		err = icoImage.Close()
 		if err != nil {
 			return "", err
 		}
+		if runtime.GOOS == "windows" {
+			cmdCandle := exec.Command("candle", fmt.Sprintf("%s.wxs", packageName))
+			cmdCandle.Dir = tmpPath
+			cmdCandle.Stdout = os.Stdout
+			cmdCandle.Stderr = os.Stderr
+			err = cmdCandle.Run()
+			if err != nil {
+				return "", err
+			}
+			cmdLight := exec.Command("light", fmt.Sprintf("%s.wixobj", packageName), "-sval")
+			cmdLight.Dir = tmpPath
+			cmdLight.Stdout = os.Stdout
+			cmdLight.Stderr = os.Stderr
+			err = cmdLight.Run()
+			if err != nil {
+				return "", err
+			}
+			err = os.Rename(filepath.Join(tmpPath, fmt.Sprintf("%s.msi", packageName)), filepath.Join(tmpPath, outputFileName))
+			if err != nil {
+				return "", err
+			}
+		} else if runtime.GOOS == "linux" {
+			cmdWixl := exec.Command("wixl", "-v", fmt.Sprintf("%s.wxs", packageName), "-o", outputFileName)
+			cmdWixl.Dir = tmpPath
+			cmdWixl.Stdout = os.Stdout
+			cmdWixl.Stderr = os.Stderr
+			err = cmdWixl.Run()
+			if err != nil {
+				return "", err
+			}
+		}
 		return outputFileName, nil
 	},
 	requiredTools: map[string][]string{
-		"linux": {"convert", "wixl"},
+		"windows": {"candle", "light"},
+		"linux":   {"wixl"},
+	},
+	generateInitFiles: func(packageName, path string) {
+		b := make([]byte, 16)
+		_, err := rand.Read(b)
+		if err != nil {
+			log.Errorf("Failed to generate GUID: %v", err)
+			os.Exit(1)
+		}
+		upgradeCode := strings.ToUpper(fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:]))
+		err = ioutil.WriteFile(filepath.Join(path, "upgrade-code.txt"), []byte(fmt.Sprintf("%s\n# This GUID is your upgrade code and ensures that you can properly update your app.\n# Don't change it.", upgradeCode)), 0755)
+		if err != nil {
+			log.Errorf("Failed to create `upgrade-code.txt` file: %v", err)
+			os.Exit(1)
+		}
+	},
+	updateTemplateData: func(packageName, path string) map[string]string {
+		data, err := ioutil.ReadFile(filepath.Join(path, "upgrade-code.txt"))
+		if err != nil {
+			log.Errorf("Failed to read `upgrade-code.txt`: %v", err)
+		}
+		guid := strings.Split(string(data), "\n")[0]
+		return map[string]string{
+			"upgradeCode":   guid,
+			"fileSeparator": string(filepath.Separator),
```

Comment:
Maybe `pathSeparator` would be a better name because it separates parts of the path, not the file.